### DISCUSSION TITLE
Partitions description (as part of PilotDescription)

### DISCUSSION THREAD

**Discussion by mtitov** [radical-cybertools/radical.pilot#2321] [General]
Andre's proposal
```python
pd = rp.PilotDescription({
    'resource': 'ornl.summit',
    'partitions': [{
        'name'  : 'prte',
        'config': 'agent_prte',
        'nodes' : 1024,
    },{
        'name'  : 'raptor',
        'config': 'agent_raptor',
        'nodes' : 8}
    ]})
```
Mikhail's updates
```python
pd = rp.PilotDescription({
    'resource': 'ornl.summit',
    'partitions': {
        'prte': {
            'config': 'agent_prte',
            'nodes' : 1024,
            'count' : 4  # number of DVMs, thus each DVM will handle up to 256 nodes
        },
        'raptor': {
            'config': 'agent_raptor',
            'nodes' : 8
        }
    })
```

**Comment by andre-merzky** [comment]
I like your's better, but with a change to the `count`: that is specific of the agent config:
```py
pd = rp.PilotDescription({
    'resource': 'ornl.summit',
    'partitions': {
        'prte': {
            'config': {
               'base': 'agent_prte',
               'dvm_count': 4},    # number of DVMs, thus each DVM will handle up to 256 nodes
            'nodes' : 1024,
        },
        'raptor': {
            'config': 'agent_raptor',
            'nodes' : 8
        }
    })
```

Otherwise we pollute the partition name space with keys which are not understood by all configurations...  That `base` thing is probably not great, but you get the idea :-)  That can be refined once the config splitup is finally done...

**Reply by mtitov** [reply]
Agree, but I think would be better to have a special key for config-related data (or even move all non-config name parameters there, including `nodes`)
```python
'prte': {
    'config'     : 'agent_prte',
    'config_data': {'dvm_count': 4},
    'nodes'      : 1024
}
```

**Reply by andre-merzky** [reply]
yeah, that works for me.

**Reply by mturilli** [reply]
I would prefer to keep it as simple (i.e., in this case as less nested) as possible. Instead of `'config'     : 'agent_prte'` would `'config_agent'     : 'agent_prte'` work? We could also use `'config_keys'` instead of the more generic `'config_data'`?

**Reply by andre-merzky** [reply]
I wonder if that is actually too much detail also: the API has otherwise no notion of an agent, or of a DVM.  Maybe leave the items generic:

```py
pd = rp.PIlotDescription()
pd.resource = 'ornl.summit'
pd.partitions = {
    'large': {
        'resource_config': 'prte',
        'nodes'          : 1024}, 
    'small': {
        'resource_config': 'default',
        'nodes'          : 16}, 
}
```

The point being: `resource_config` would basically reference `ornl.summit_prte`, and config options would remain in the config files by default.  If we want to overwrite configs, we would use the following format:
```py
pd = rp.PIlotDescription()
pd.resource = 'ornl.summit'
pd.partitions = {
    'large': {
        'resource_config': {
            'base'     : 'ornl.summi_prte',
            'dmv_count': 4
            },
        'nodes': 1024}, 
    'small': {
        'resource_config': 'default',
        'nodes'          : 16}, 
}
```

So instead of _referencing_ a config for that partition we would _pass_ a config.  But that second  form should not be documented for end user IMHO, just like we don't encourage end users to fiddle with resource config details all the time.

**Reply by mtitov** [reply]
I would go for to keep parameters types as they are (not to overcomplicated the structure), config name and number of nodes always would be defined here, and if some parameters should be overwritten then to have a separate key for that data. `config_data` name fits that because it would contain key-value pairs, if to have it as `config_keys` that could be considered as just list of keys. Parameter name could be as an action/verb, e.g. `'update': {'dvm_count': 5}`

**Reply by mtitov** [reply]
and one more thing, we didn't state (at least here) what will be in config part for partitions, I would think that some reorganization in resource config would be needed then - e.g., to have the same description approach as for `access_schema`, in a way to group together keys, which could be a description for partitions, example
```
{
    "summit": {
        ...
        "partition_default"                   : "jsrun",
        "partitions": {
            "jsrun": {
                "agent_config"                : "default",
                "agent_scheduler"             : "CONTINUOUS",
                "agent_launch_method"         : "JSRUN",
                "task_launch_method"          : "JSRUN",
                "mpi_launch_method"           : "JSRUN",
                "pre_exec"                    : [...]
            },
            "prte": {
                "agent_config"                : "prte_sa",
                "agent_scheduler"             : "CONTINUOUS_COLO",
                "agent_launch_method"         : "SSH",
                "task_launch_method"          : "PRTE",
                "mpi_launch_method"           : "PRTE",
                "dvm_count"                   : 4,
                "pre_exec"                    : [...]
            }
        }
    }
}
```

and then have the following in code
```python
pd = rp.PilotDescription()
pd.resource = 'ornl.summit'
pd.partitions = {
    'jsrun': {
        'nodes'  : 1024}, 
    'prte': {
        'update' : {'dvm_count': 10},
        'nodes'  : 16}, 
}
```
In case of a separate config for partitions, then corresponding names would be provided
